src/tracer.js

```diff
@@ -0,0 +1,218 @@
+                fields.childOf.traceId,
+                randomId,
+                fields.childOf.spanId,
+                fields.childOf.flags,
+                fields.childOf.baggage
+            );
+        } else {
+            spanContext = new SpanContext(
+                randomId,
+                randomId,
+                null,
+                flags,
+                this._tracerTags
+            );
+        }
+
+        // todo(oibe) in the new model how will the instrumentation specify the peer for
+        // a client, and server span?
+        // todo(oibe) also make sure we can remove applyBeginOptions
+        // todo(oibe) support startTime?
+        // todo(oibe) support tags here?
+        return this._createSpan(fields.operationName, spanContext, isClient, fields.peer);
+    }
+
+    /**
+     * Saves the span context into the carrier object for various formats, and encoders.
+     *
+     * @param  {SpanContext} spanContext - the SpanContext to inject into the
+     *         carrier object. As a convenience, a Span instance may be passed
+     *         in instead (in which case its .context() is used for the
+     *         inject()).
+     * @param  {string} format - the format of the carrier.
+     * @param  {any} carrier - see the documentation for the chosen `format`
+     *         for a description of the carrier object.
+     **/
+    inject(spanContext: SpanContext, format: string, carrier: any): void {
+        let injector = this._injectors[format];
+        if (!injector) {
+            throw `Unsupported format: ${format}`;
+        }
+
+        injector.inject(spanContext, carrier);
+    }
+
+    /**
+    * Responsible for extracting a span context from various serialized formats.
+    *
+    * @param  {string} format - the format of the carrier.
+    * @param  {any} carrier - the type of the carrier object is determined by
+    *         the format.
+    * @return {SpanContext}
+    *         The extracted SpanContext, or null if no such SpanContext could
+    *         be found in `carrier`
+    */
+    extract(format: string, carrier: any): SpanContext {
+        let extractor = this._extractors[format];
+        if (!extractor) {
+            throw `Unsupported format: ${format}`;
+        }
+
+        return extractor.extract(carrier);
+    }
+
+    /**
+     * Closes the tracer, flushes spans, and executes any callbacks if necessary.
+     *
+     * @param {Function} [callback] - a callback that runs after the tracer has been closed.
+     **/
+    close(callback: Function): void {
+        var reporter = this._reporter;
+        this._reporter = NoopReporter();
+        reporter.close(() => {
+            this._sampler.close(callback);
+        });
+    }
+
+    /**
+     * Writes spans to the local agent, and executes a callback if necessary.
+     *
+     * @param {Function} [callback] - a callback that runs after spans have been flushed.
+     **/
+    flush(callback: Function): void {
```

Comment:
ah, interesting. That's not present in OT for any other languages.

Fine, then we need flush on the top-level reporter interface.

But we still need close, which is slightly different - it assumes flush as well, but also does clean-up like timers etc.